File: problems/0146/solution.cpp
#include <unordered_map>
using namespace std;

struct Node {
  int key;
  int val;
  Node* prev;
  Node* next;
  Node(int key, int val) : key(key), val(val) {}
};

class LRUCache {
  int capacity;
  unordered_map<int, Node*> mp;
  Node* head;
  Node* tail;
  void removeNode(Node* node) {
    node->prev->next = node->next;
    node->next->prev = node->prev;
  }
  void addNode(Node* node) {
    node->prev = head;
    node->next = head->next;
    head->next = node;
    node->next->prev = node;
  }

 public:
  LRUCache(int capacity) : capacity(capacity) {
    head = new Node(-1, 0);
    tail = new Node(-1, 0);
    head->next = tail;
    tail->prev = head;
  }

  int get(int key) {
    if (mp.count(key)) {
      removeNode(mp[key]);
      addNode(mp[key]);
      return mp[key]->val;
    }
    return -1;
  }

  void put(int key, int value) {
    if (get(key) != -1) {
      mp[key]->val = value;
      return;
    }
    Node* cache = new Node(key, value);
    addNode(cache);
    mp[key] = cache;
    if (capacity) {
      capacity--;
    } else {
      int oldest_key = tail->prev->key;
      removeNode(mp[oldest_key]);
      delete mp[oldest_key];
      mp.erase(oldest_key);
    }
  }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: problems/0146/solution.py
class Node:
    def __init__(self, key, val):
        self.key: int = key
        self.val: int = val
        self.next: Node = None
        self.prev: Node = None


class LRUCache:
    def __init__(self, capacity: int):
        self.capacity: int = capacity
        self.mp: dict[int, Node] = {}
        self.head: Node = Node(-1, 0)
        self.tail: Node = Node(-1, 0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def removeNode(self, node: Node):
        node.prev.next = node.next
        node.next.prev = node.prev

    def addNode(self, node: Node):
        node.prev = self.head
        node.next = self.head.next
        self.head.next = node
        node.next.prev = node

    def get(self, key: int) -> int:
        if key in self.mp:
            self.removeNode(self.mp[key])
            self.addNode(self.mp[key])
            return self.mp[key].val
        return -1

    def put(self, key: int, value: int) -> None:
        if self.get(key) != -1:
            self.mp[key].val = value
            return
        cache = Node(key, value)
        self.addNode(cache)
        self.mp[key] = cache
        if self.capacity:
            self.capacity -= 1
        else:
            oldest_key = self.tail.prev.key
            self.removeNode(self.mp[oldest_key])
            self.mp.pop(oldest_key)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
